scripts/Stretch.py

```diff
@@ -0,0 +1,357 @@
+
+    ##  Convenience function that finds the value in a line of g-code.
+    #   When requesting key = x from line "G1 X100" the value 100 is returned.
+    #   It is a copy of Stript's method, so it is no DontRepeatYourself, but
+    #   I split the class into setup part (Stretch) and execution part (Strecher)
+    #   and only the setup part inherits from Script
+    def getValue(self, line, key, default = None):
+        if not key in line or (";" in line and line.find(key) > line.find(";")):
+            return default
+        sub_part = line[line.find(key) + 1:]
+        m = re.search("^-?[0-9]+\.?[0-9]*", sub_part)
+        if m is None:
+            return default
+        try:
+            return float(m.group(0))
+        except:
+            return default
+
+    def execute(self,data):
+        Logger.log("d","Post stretch with line width=" + str(self.line_width) + "mm and stretch=" + str(self.stretch)+ "mm")
+        retdata = []
+        layerSteps = []
+        current_x = 0.
+        current_y = 0.
+        current_z = 0.
+        current_e = 0.
+        current_f = 0.
+        layer_z = 0.
+        for layer in data:
+            lines = layer.rstrip("\n").split("\n")
+            for line in lines:
+                comment = ""
+                if line.find(";") >= 0:
+                    comment = line[line.find(";"):]
+                if self.getValue(line,"G") == 0:
+                    current_x = self.getValue(line,"X",current_x)
+                    current_y = self.getValue(line,"Y",current_y)
+                    current_z = self.getValue(line,"Z",current_z)
+                    current_e = self.getValue(line,"E",current_e)
+                    current_f = self.getValue(line,"F",current_f)
+                    onestep = GCodeStep(0,current_x,current_y,current_z,current_e,current_f,comment)
+                elif self.getValue(line,"G") == 1:
+                    current_x = self.getValue(line,"X",current_x)
+                    current_y = self.getValue(line,"Y",current_y)
+                    current_z = self.getValue(line,"Z",current_z)
+                    current_e = self.getValue(line,"E",current_e)
+                    current_f = self.getValue(line,"F",current_f)
+                    onestep = GCodeStep(1,current_x,current_y,current_z,current_e,current_f,comment)
+                elif self.getValue(line,"G") == 92:
+                    current_x = self.getValue(line,"X",current_x)
+                    current_y = self.getValue(line,"Y",current_y)
+                    current_z = self.getValue(line,"Z",current_z)
+                    current_e = self.getValue(line,"E",current_e)
+                    current_f = self.getValue(line,"F",current_f)
+                    onestep = GCodeStep(-1,current_x,current_y,current_z,current_e,current_f,comment)
+                else:
+                    onestep = GCodeStep(-1,current_x,current_y,current_z,current_e,current_f,line);
+                if current_z != layer_z:
```

Comment:
Anyway, as you have certainly noticed, I have a poor knowledge of python and did a rough translation from French C++ into English Python. I will try to enhance ;-)